test_util/aws.py

```diff
@@ -74,49 +75,23 @@ def __init__(self, stack_name, boto_wrapper):
         self.boto_wrapper = boto_wrapper
         self.stack = self.boto_wrapper.resource('cloudformation').Stack(stack_name)
 
-    def wait_for_status_change(self, state_1, state_2, timeout=3600):
-        """Note: Do not use boto waiter class, it has very poor error handling
-        and will raise an exception when the rate limit is hit whereas botoclient
-        methods will simply sleep and retry up to 4 times. After that a ClientError
-        is raised, at which point the poll interval backs off
+    def wait_for_status_change(self, state_1, state_2, timeout=60 * 60):
         """
-        stack_states = [
-            'CREATE_IN_PROGRESS', 'CREATE_FAILED', 'CREATE_COMPLETE',
-            'ROLLBACK_IN_PROGRESS', 'ROLLBACK_FAILED', 'ROLLBACK_COMPLETE',
-            'DELETE_IN_PROGRESS', 'DELETE_FAILED', 'DELETE_COMPLETE',
-            'UPDATE_IN_PROGRESS', 'UPDATE_COMPLETE_CLEANUP_IN_PROGRESS',
-            'UPDATE_COMPLETE', 'UPDATE_ROLLBACK_IN_PROGRESS',
-            'UPDATE_ROLLBACK_FAILED', 'UPDATE_ROLLBACK_COMPLETE',
-            'UPDATE_ROLLBACK_COMPLETE_CLEANUP_IN_PROGRESS']
-        state_args = set([state_1, state_2])
-        assert state_args.issubset(stack_states), 'Invalid state(s): {}. states must be one of: {}'.format(
-            repr(state_args.difference(stack_states)), repr(stack_states))
-        logging.info('Waiting for status to change from {} to {}'.format(state_1, state_2))
-
-        logging.info('Sleeping for {} minutes before polling'.format(AWS_WAIT_BEFORE_POLL_MIN))
+        Stacks can have one of the following statuses:
+        CREATE_IN_PROGRESS, CREATE_FAILED, CREATE_COMPLETE
+        ROLLBACK_IN_PROGRESS, ROLLBACK_FAILED, ROLLBACK_COMPLETE
+        DELETE_IN_PROGRESS, DELETE_FAILED, DELETE_COMPLETE
+        UPDATE_IN_PROGRESS, UPDATE_COMPLETE_CLEANUP_IN_PROGRESS
+        UPDATE_COMPLETE, UPDATE_ROLLBACK_IN_PROGRESS
+        UPDATE_ROLLBACK_FAILED, UPDATE_ROLLBACK_COMPLETE
+        UPDATE_ROLLBACK_COMPLETE_CLEANUP_IN_PROGRESS
+        """
+        logger.info('Waiting for status to change from {} to {}'.format(state_1, state_2))
+        logger.info('Sleeping for {} minutes before polling'.format(AWS_WAIT_BEFORE_POLL_MIN))
         time.sleep(60 * AWS_WAIT_BEFORE_POLL_MIN)
-        default_poll_interval = 10
 
-        @retrying.retry(stop_max_delay=timeout * 1000,
-                        retry_on_result=lambda res: res is False,
-                        retry_on_exception=lambda ex: False)
-        def wait_and_backoff():
-            nonlocal default_poll_interval
-            try:
-                self._wait_loop(state_1, state_2, default_poll_interval)
-            except ClientError as e:
-                if e.response['Error']['Code'] == 'Throttling':
-                    logging.warn('AWS Client raised a throttling error; increasing poll interval...')
-                    default_poll_interval += 10
```

Comment:
The reason I decided on this was that I wanted to make decorator that represented "try this function and if it gets this error, wait, and then try again, with waits getting successively longer. That way I could wrap functions that didnt have the concept of repeatedly calling polling AWS.

The effect that this has for functions that repeatedly poll is that we now just poll at a constant interval. back off. poll at the same interval. back off for even longer. repeat. My hand-wavy reasoning is that geometrically increasing backoffs (wherein no polling occurs) are roughly equivalent (in terms of overall API use) to a polling interval that increases linearly after failure